Support iscsi volume attach and detach

mtanino opened the pull request:
**What this PR does / why we need it**:

Support iSCSI volume attach and detach to enforce RWO access

**Which issue this PR fixes** : fixes #48953

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
NONE
```

mtanino commented:
/sig storage

rootfs commented:
/approve

rootfs commented:
@deads2k can you approve? thanks

rootfs commented:
cc / @saad-ali @jingxu97 @msau42 for `WaitForAttach` interface change.

rootfs commented:
@mtanino can you make 2 commits? one for `WaitForAttach` refactoring and the other for iscsi refactoring. thanks

Commit:
WaitForAttach refactoring for iSCSI attacher/detacher

This change is prerequisite for implementing iSCSI attacher
and detacher.

In order to use chap authentication at iSCSI plugin after
implementing attacher and detacher, secret is needed at
AttachDisk() which is called from WaitForAttach().
To obtain secret, pod information is required, but
WaitForAttach() doesn't pass pod information inside.

This patch adds 'pod' as an argument of WaitForAttach()
and adds changes to drivers who implements WaitForAttach().

Fixes #48953

Commit:
Support iscsi volume attach and detach

Fixes #48953

mtanino commented:
@rootfs 
I separated commits for two as you mentioned.
The fix for attacher/detacher is done but I'm still testing iSCSI piece to avoid degradation.
I would like to get feedback for WaitForAttach interface change at first.

mtanino commented:
/retest

mtanino commented:
pkg/volume/iscsi/BUILD is updated to solve bazel error.

jingxu97 commented:
@mtanino could you add some comments with more details about the need for waitForAttach() change? Thanks!

mtanino commented:
@jingxu97 @rootfs 
I've added background of WaitForAttach refactoring to the commit message. Thanks.

deads2k commented:
controller manager changes are ok 
/approve

mtanino commented:
/retest

mtanino commented:
@jingxu97 @msau42 @saad-ali 
Could you give feedbacks for the WaitForAttach interface change?

Thanks.

rootfs commented:
@mtanino is it still WIP?

mtanino commented:
@rootfs 
Patch is all set. I need some more tests.

rootfs commented on pkg/volume/iscsi/attacher.go:
```diff
@@ -0,0 +1,205 @@
+		options = append(options, "ro")
+	}
+	if notMnt {
+		diskMounter := &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()}
```
use the new `exec.Run` method

mtanino replied:
ok. This will conflict with @jsafrane 's ongoing fix.

jingxu97 commented on pkg/volume/iscsi/attacher.go:
```diff
@@ -0,0 +1,205 @@
+	return mount.GetMountRefs(mounter, deviceMountPath)
+}
+
+func (attacher *iscsiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
```
why this function is doing nothing?

mtanino replied:
@jingxu97 
As for the iSCSI plugin, we have to call volume attach operations(iscsiadm commands) on the kubelet node.
I think this Attach() method will be executed on the controller node. iSCSI plugin doesn't need anything to do on the controller node. That's why this is doing nothing.
Main attachment logic is in the WaitForAttach().

jingxu97 commented on pkg/volume/iscsi/attacher.go:
```diff
@@ -0,0 +1,205 @@
+func (attacher *iscsiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.NodeName) (map[*volume.Spec]bool, error) {
+	volumesAttachedCheck := make(map[*volume.Spec]bool)
+	for _, spec := range specs {
+		volumesAttachedCheck[spec] = true
```
all checks return true?

mtanino replied:
@jingxu97 
Since iSCSI plugin doesn't have a way to confirm whether the volume is attached or not from outside of the kubelet node, we think this method needs to return true always. This is similar to the FC plugin attacher/detacher implementation.

jingxu97 commented:
@mtanino many functions in the new attacher for iScsi look strange to me. Do you have some docs or information about the changes? Thanks!

mtanino commented:
@jingxu97 @rootfs 
I think the point is iSCSI plugin and FC plugin doesn't have cloud provider.
Therefore, we can't implement attacher/detacher straight forward like other plugins such as GCE, AWS, etc.

jingxu97 commented:
@mtanino thanks for the information. But I feel the function behavior of iscsi or fc would be very different from other volume plugins attacher interface. Maybe we could make reconciler determine that this type of volume requires attach invoked by kubelet instead of master so kubelet will issue attach operation. We just need to modify a little in here https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L215
@saad-ali do you have comments about this?

mtanino commented:
@jingxu97 
I see. If we can enforce RWO access via reconciler change for specific plugins, I think it's fine.

@rootfs 
Any comments, concerns?

rootfs commented:
@jingxu97 the idea of implementing iscsi attach/detach is to use [isMultiAttachForbidden](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L139) to enfore RWO

mtanino commented:
/retest

mtanino commented:
/retest

rootfs commented:
/lgtm

mtanino commented:
/retest

rootfs commented:
@mtanino can you rebase? test and compile failed with latest master code.

rootfs commented:
rebase looks working

rootfs commented:
/lgtm
